### PR TITLE
Check symbol table mappings

### PR DESCRIPTION
**Pull request by chrisr-diffblue**
This PR builds upon/depends upon #3123 (merged). It adds some basic sanity checking of the symbol table structure, namely that symbols are correctly keyed, and that `base_name` and `module` maps are in sync with the symbols.
 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by martin-cs** [comment]
Do you want reviews before #3123 is merged?

**Comment by chrisr-diffblue** [comment]
Happy to receive reviews before #3123 - the checks added in this PR are independent of #3123, it just uses that PR for its hooks.

**Review by martin-cs — APPROVED** [review]
I have only reviewed the last commit; so approval is for that, not the whole stack.

File: src/util/symbol_table.cpp
Comment (xbauch) [inline src/util/symbol_table.cpp]:
```    
      bool map_matches =
        std::find_if(
          symbol_base_map.begin(),
          symbol_base_map.end(),
          [&](const typename symbol_base_mapt::value_type &pair) {
            return pair.first == symbol.base_name && pair.second == symbol_key;
          }) != symbol_base_map.end();
```
Doesn't the while loop actually go through all the elements past the first `find` match?

Comment (chrisr-diffblue) [reply src/util/symbol_table.cpp]: The `!map_matches` condition in the while loop will terminate the loop at the first match, but I agree that your suggested use of the std library is a better way to go - I'll update, thanks :-)

Comment (chrisr-diffblue) [reply src/util/symbol_table.cpp]: Actually, using `equal_range` is probably better....

Comment (xbauch) [reply src/util/symbol_table.cpp]: Yes `equal_range` would be logarithmic, unlike the above.

**Review by NlightNFotis — APPROVED** [review]
LGTM

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0f18b68965fdd680bd808390dbb2cd6e8153f66d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91054169

**Review by tautschnig — APPROVED** [review]

**Pull request merged**
